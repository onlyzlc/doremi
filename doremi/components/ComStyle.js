import { StyleSheet } from "react-native";

export const Colors = {
  light: "rgb(231, 238, 252)",
  main: "rgb(65, 35, 68)",
  cantrol: "rgb(65, 35, 68)",
  foreground: "rgb(255, 255, 255)",
  secondary: "rgb(255, 192, 60)",
  secondText: "#8B668E",
  secBackground: "#B1C4EC",
};

export const Fonts = {
  h1: {
    fontSize: 32,
    fontWeight: 600,
    textAlign: "left",
    // color: Colors.main,
  },
  h2: {
    fontSize: 24,
    fontWeight: 400,
    textAlign: "left",
    // color: Colors.main,
  },
  h3: {
    fontSize: 16,
    fontWeight: 400,
    textAlign: "left",
    // color: Colors.main,
  },
  main: {
    fontSize: 16,
    fontWeight: 400,
    textAlign: "left",
    // color: Colors.main,
  },
  secondary: {
    fontSize: 14,
    fontWeight: 400,
    textAlign: "left",
    // color: Colors.secondText,
  },
};

// 统一的容器两侧边距
const spacing1 = 16;

export const Styles = StyleSheet.create({
  screen: {
    flex: 1,
    backgroundColor: Colors.light,
  },
  screenBackground: {
    flex: 1,
  },
  titleBar: {
    width: "100%",
    display: "flex",
    flexDirection: "row",
    paddingHorizontal: spacing1,
    justifyContent: "spaceBetween",
    alignItems: "center",
  },
  section: {
    padding: spacing1,
  },
  sectionTitle: {
    paddingVertical: spacing1,
  },
  card: {
    borderRadius: 8,
    justifyContent: "spaceBetween",
    padding: spacing1,
    gap: 8,
    backgroundColor: Colors.foreground,
  },
});
