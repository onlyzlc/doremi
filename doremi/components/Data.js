export const solfa = ["do", "re", "mi", "fa", "sol", "la", "si"];
export const homophone = ["哆", "来", "咪", "发", "所", "啦", "西"];
export const song1 = `N: (12) 3 4 | 5678 |
          Lw: Do, a deer, a fe-male deer. Re, a drop of gol-den sun.
          ---
          N: 4321 | 1234| 
          Lw: Mi, a name I call my-self. Fa, a long long way to run.`;
