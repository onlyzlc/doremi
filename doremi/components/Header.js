/* eslint-disable react/prop-types */
import { React } from "react";
import { StyleSheet, Text } from "react-native";

export const H1 = ({ title, ...props }) => {
  return <Text style={[styles.header, styles.h1, props.style]}>{title}</Text>;
};
export const H2 = ({ title, ...props }) => {
  return <Text style={[styles.header, styles.h2, props.style]}>{title}</Text>;
};
export const H3 = ({ title, ...props }) => {
  return <Text style={[styles.header, styles.h3, props.style]}>{title}</Text>;
};

const styles = StyleSheet.create({
  header: {
    color: "rgb(51, 51, 51)",
    fontWeight: 500,
    textAlign: "left",
  },
  h1: {
    fontSize: 32,
  },
  h2: {
    fontSize: 24,
  },
  h3: {
    fontSize: 16,
  },
});
