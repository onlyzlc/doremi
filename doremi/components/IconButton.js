/* eslint-disable react/prop-types */
import { React } from "react";
import { StyleSheet, View, Pressable, Text } from "react-native";
import MaterialIcons from "@expo/vector-icons/MaterialIcons";

export default function IconButton({
  onPress,
  name,
  label,
  size,
  disabled = false,
}) {
  return (
    <View style={[{ width: size }]}>
      <Pressable onPress={onPress} style={[styles.button]} disabled={disabled}>
        <MaterialIcons name={name} size={size / 2} style={styles.icon} />
        {label && <Text style={styles.iconLabel}>{label}</Text>}
      </Pressable>
    </View>
  );
}

const styles = StyleSheet.create({
  button: {
    justifyContent: "center",
  },
  icon: {
    textAlign: "center",
    color: "#000",
    padding: 8,
  },
  iconLabel: {
    color: "#000",
    textAlign: "center",
  },
});
