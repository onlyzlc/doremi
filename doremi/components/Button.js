/* eslint-disable react/prop-types */
import { React } from "react";
import { StyleSheet, View, Pressable, Text } from "react-native";
import { Colors } from "./ComStyle";

export default function Button({ title, theme, onPress, disabled, ...props }) {
  if (theme == "primary") {
    return (
      <View style={[styles.buttonContainer]}>
        <Pressable
          style={[styles.button, { backgroundColor: "#0A59F7" }]}
          onPress={onPress}>
          <Text style={[styles.buttonLabel, { color: "#fff" }]}> {title}</Text>
        </Pressable>
      </View>
    );
  }
  return (
    <View style={styles.buttonContainer}>
      <Pressable style={styles.button} onPress={onPress} {...props}>
        <Text style={styles.buttonLabel}>{title}</Text>
      </Pressable>
    </View>
  );
}

const styles = StyleSheet.create({
  buttonContainer: {
    minWidth: 120,
    minHeight: 60,
  },
  button: {
    borderRadius: 40,
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
    flexDirection: "row",
    backgroundColor: Colors.secBackground,
  },
  buttonLabel: {
    color: Colors.main,
    fontSize: 16,
    textAlign: "center",
  },
  buttonIcon: {
    paddingLeft: 8,
  },
});
