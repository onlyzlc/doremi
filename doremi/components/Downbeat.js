// 起拍指示符
import { View, StyleSheet, Pressable } from "react-native";
import { Colors } from "./ComStyle";
import Animated, { FadeIn, FadeOut } from "react-native-reanimated";
import React, { useEffect, useState } from "react";

function Dot({ delay = 0 }) {
  const [show, setShow] = useState(true);
  useEffect(() => {
    const timer = setTimeout(() => setShow(false), delay);
    return () => {
      clearTimeout(timer);
      setShow(true);
    };
  }, [delay]);
  return (
    <View>
      {show && (
        <Animated.View style={styles.dot} exiting={FadeOut}></Animated.View>
      )}
    </View>
  );
}

export default function DownBeat({ tempo = 60, signature = 4 }) {
  // 每拍时长(毫秒)
  const dur = 60000 / tempo;
  // 生成一个序列再渲染的方式
  // const dots = Array.from({ length: signature }, (v, i) => i);
  // return (
  //   <View style={styles.container}>
  //     {dots.map((i) => (
  //       <Dot delay={dur * (signature - i)} key={i} />
  //     ))}
  //   </View>
  // );
  const dots = [];
  for (let i = 0; i < signature; i++) {
    dots.push(<Dot delay={dur * (signature - i)} key={i} />);
  }
  return <View style={styles.container}>{dots}</View>;
}

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    gap: 32,
  },
  dot: {
    width: 12,
    height: 12,
    borderRadius: 12,
    backgroundColor: Colors.secondary,
  },
});
