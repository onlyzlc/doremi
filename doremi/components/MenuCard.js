import { View, Text, Pressable } from "react-native";
import { Styles, Fonts, Colors } from "./ComStyle";
import { React } from "react";

export default function MenuCard({ title = "", onPress, children }) {
  return (
    <Pressable onPress={onPress} style={[Styles.card]}>
      <Text style={Fonts.h3}>{title}</Text>
      {children}
    </Pressable>
  );
}
