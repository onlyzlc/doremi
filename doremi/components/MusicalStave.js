import { StyleSheet, View, Text } from "react-native";
import { React } from "react";
import { Colors } from "./ComStyle";
import { parse } from "./Parse";
import { type } from "@testing-library/react-native/build/user-event/type";

// times:  1=1拍, 0.5=半拍...
export function Note({ note = 0, status = "", times = 1 }) {
  const noteStyles = StyleSheet.create({
    default: {
      fontSize: 16,
      color: Colors.main,
    },
    zoom: {
      fontSize: 24,
    },
    wrong: {
      color: Colors.foreground,
    },
  });

  const noteBoxStyles = StyleSheet.create({
    default: {
      // width: 30,
      height: 40,
      borderRadius: 4,
    },
    zoom: {
      borderStyle: "2px solid rgb(255, 192, 60)",
    },
    wrong: {
      color: Colors.red,
    },
  });

  return (
    <View
      style={[
        noteBoxStyles.default,
        status && noteBoxStyles[status],
        { width: times * 20 },
      ]}>
      <Text style={[noteStyles.default, status && noteStyles[status]]}>
        {note}
      </Text>
    </View>
  );
}
// {
//   properties: "",
//   body: [
//     { //chip
//       N: [[小节],[x,x,x,[半拍]]],
//       lc1: [],
//       lc2: [],
//     },
//   ],
// };
// b1:1拍  b2:半拍  b4:四分之一拍  b8:八分之一拍

let noteIndex = 0;
export function BarNotes({ barData, times = 1 }) {
  if (typeof barData !== "object") return;
  const listItems = barData.map((beat) => {
    if (typeof beat === "string") {
      return <Note key={"n" + noteIndex++} note={beat} duration={times} />;
    } else if (typeof beat === "object") {
      return (
        <BarNotes
          key={"n" + noteIndex + "." + times}
          barData={beat}
          times={times * 0.5}
        />
      );
    }
  });
  return <>{listItems}</>;
}

function BarLine() {
  return <Text>|</Text>;
}

let barIndex = 0;
// 输入: [ [ '1', '2', '3', '4' ], [ '1', '2', '3', '4' ] ]
export function Chip({ chipData }) {
  console.log("Chip输入:", chipData);
  if (typeof chipData !== "object") return;
  const listItems = chipData.map((bar) => (
    <View style={styles.bar} key={"b" + barIndex++}>
      <BarNotes barData={bar} />
      <BarLine />
    </View>
  ));
  return <View style={styles.noteLine}>{listItems}</View>;
}

let chipIndex = 0;
export default function MusicalStave({ staveDoc }) {
  const stave = parse(staveDoc);
  if (typeof stave !== "object") return;
  const listItems = stave.body.map((item) => {
    return item.N && <Chip key={"c" + chipIndex++} chipData={item.N} />;
  });
  return <View>{listItems}</View>;
}

const styles = StyleSheet.create({
  bar: {
    flexDirection: "row",
  },
  noteLine: {
    flexDirection: "row",
  },
});
