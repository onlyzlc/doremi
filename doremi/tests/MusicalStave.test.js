/* eslint-disable react/react-in-jsx-scope */
import { render } from "@testing-library/react-native";
import MusicalStave, { Chip, Bar } from "../components/MusicalStave";
import { song1 } from "../components/Data";

// test("渲染小节", () => {
//   const data = ["1e", "b2"];
//   const tree = render(<Bar barData={data} />).toJSON();
//   expect(tree).toMatchSnapshot();
// });
// test("渲染碎片", () => {
//   const data = [
//     ["1", "2"],
//     ["3", "4"],
//   ];
//   const tree = render(<Chip chipData={data} />).toJSON();
//   expect(tree).toMatchSnapshot();
// });
test("渲染简谱", () => {
  const tree = render(<MusicalStave staveDoc={song1} />).toJSON();
  expect(tree).toMatchSnapshot();
});
