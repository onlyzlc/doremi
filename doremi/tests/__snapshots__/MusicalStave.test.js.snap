// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`渲染简谱 1`] = `
<View>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          1
        </Text>
      </View>
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          2
        </Text>
      </View>
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          3
        </Text>
      </View>
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          4
        </Text>
      </View>
      <Text>
        |
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          5
        </Text>
      </View>
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          6
        </Text>
      </View>
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          7
        </Text>
      </View>
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          8
        </Text>
      </View>
      <Text>
        |
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          4
        </Text>
      </View>
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          3
        </Text>
      </View>
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          2
        </Text>
      </View>
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          1
        </Text>
      </View>
      <Text>
        |
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          1
        </Text>
      </View>
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          2
        </Text>
      </View>
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          3
        </Text>
      </View>
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 40,
            },
            "",
            {
              "width": 20,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(65, 35, 68)",
                "fontSize": 16,
              },
              "",
            ]
          }
        >
          4
        </Text>
      </View>
      <Text>
        |
      </Text>
    </View>
  </View>
</View>
`;
